Source/WebCore/style/values/inline/StyleLineHeight.cpp

```diff
@@ -0,0 +1,150 @@
+{
+    RefPtr primitiveValue = requiredDowncast<CSSPrimitiveValue>(state, value);
+    if (!primitiveValue)
+        return CSS::Keyword::Normal { };
+
+    return this->operator()(state, *primitiveValue, multiplier);
+}
+
+auto CSSValueConversion<LineHeight>::operator()(BuilderState& state, const CSSPrimitiveValue& primitiveValue, float multiplier) -> LineHeight
+{
+    auto valueID = primitiveValue.valueID();
+    if (valueID == CSSValueNormal || CSSPropertyParserHelpers::isSystemFontShorthand(valueID))
+        return CSS::Keyword::Normal { };
+
+    auto conversionData = state
+        .cssToLengthConversionData()
+        .copyForLineHeight(zoomWithTextZoomFactor(state));
+
+    if (primitiveValue.isLength() || primitiveValue.isCalculatedPercentageWithLength()) {
+        double fixedValue = 0;
+        if (primitiveValue.isLength())
+            fixedValue = primitiveValue.resolveAsLength(conversionData);
+        else
+            fixedValue = primitiveValue.protectedCssCalcValue()->createCalculationValue(conversionData, CSSCalcSymbolTable { })->evaluate(state.style().computedFontSize());
+
+        if (multiplier != 1.0f)
+            fixedValue *= multiplier;
+
+        return LineHeight::Fixed {
+            CSS::clampToRange<LineHeight::Fixed::range, float>(fixedValue, minValueForCssLength, maxValueForCssLength)
+        };
+    }
+
+    // Line-height percentages need to inherit as if they were Fixed pixel values. In the example:
+    // <div style="font-size: 10px; line-height: 150%;"><div style="font-size: 100px;"></div></div>
+    // the inner element should have line-height of 15px. However, in this example:
+    // <div style="font-size: 10px; line-height: 1.5;"><div style="font-size: 100px;"></div></div>
+    // the inner element should have a line-height of 150px. Therefore, we map percentages to Fixed
+    // values and raw numbers to percentages.
+    if (primitiveValue.isPercentage()) {
+        // FIXME: percentage should not be restricted to an integer here.
+        return LineHeight::Fixed {
+            CSS::clampToRange<LineHeight::Fixed::range, float>((state.style().computedFontSize() * primitiveValue.resolveAsPercentage<int>(conversionData)) / 100.0, minValueForCssLength, maxValueForCssLength)
+        };
+    }
+
+    if (primitiveValue.isNumber()) {
+        return LineHeight::Percentage {
+            CSS::clampToRange<LineHeight::Percentage::range, float>(primitiveValue.resolveAsNumber(conversionData) * 100.0)
+        };
+    }
+
+    state.setCurrentPropertyInvalidAtComputedValueTime();
+    return CSS::Keyword::Normal { };
+}
+
+// MARK: - Blending
+
+auto Blending<LineHeight>::canBlend(const LineHeight& a, const LineHeight& b) -> bool
+{
+    if (a.hasSameType(b))
+        return true;
+
+    if (a.isCalculated())
+        return b.isSpecified();
+    if (b.isCalculated())
+        return a.isSpecified();
+
+    return false;
```

Comment:
Maybe a boolean expression here?
```suggestion
    return a.hasSameType(b) || (a.isCalculated() && b.isSpecified()) || (b.isCalculated() && a.isSpecified());
```